BEST_PRACTICES.md

```diff
@@ -0,0 +1,82 @@
+`node` blocks, so that it actually happens on an executor.
+* Get your flows from source control - `Jenkinsfile`s, loading
+libraries, global CPS library, you name it - but if you pull the main flow
+from SCM (i.e., Multibranch with `Jenkinsfile`s or `Workflow script from
+SCM`), be aware that you may need to whitelist a lot of method calls in
+the script security
+plugin. [JENKINS-28178](https://issues.jenkins-ci.org/browse/JENKINS-28178)
+has been created to simplify this, but you'll still want to be careful
+with that functionality due to the open security it implies.
+* `input` shouldn’t be done within a `node` block - that eats up two
+executors waiting on the `input` feedback, both the flyweight executor
+that’s used for the `input` itself and the real executor used by the
+`node` block, which won’t free up until after the `input` itself has completed.
+
+# Parallelism
+* Within `parallel` blocks, use `node` blocks to make sure you farm out to
+real nodes for your parallelized work.
+* Nested `parallel` blocks can lead to swamping your available
+executors, as each execution of the first `parallel` block calls
+multiple executions of the second `parallel` block, and so on. In
+general, think carefully about your parallelism and your available executors
+when using `parallel`.
+* The [Parallel Test Executor
+plugin](https://github.com/jenkinsci/parallel-test-executor-plugin) is
+awesome and can be immensely helpful both for distributing your test
+execution and for throttling your parallelism, since you define how
+many "buckets" your tests get dviided into.
+* Don’t put `stage`s in `parallel` blocks - that just goes *weird*,
+breaking a lot of logic in the Stage View and elsewhere. Save yourself
+the pain - don't do it!
+
+# `Jenkinsfile`s and Multibranch
+* Use: `checkout scm` to automatically checkout current revision of branch
+* Use: `$env.BRANCH_NAME` variable if you have logical difference in
+your flow between branches, i.e. to distinguish different behavior for
+production-ready branches versus sandbox or pull request branches.
+* For `Jenkinsfile`s, make sure to put `#!groovy` at the top of the
+file so that IDEs, GitHub diffs, etc properly detect the language and
+do syntax highlighting for you.
+
+# Groovy gotchas
+* Don’t have the Groovy interpreter making blocking i/o calls, i.e.,
+`HTTPClient` and the like - these can cause real problems with
+resumability, and also require a lot of explicit whitelisting of
+methods in the Script Security plugin, which is not ideal.
+* Beware `for (Foo f: foos)` loops and Groovy closure-style operators
+like `.each` and the like. They will not work right in normal Workflow
+script contexts where Workflow steps are involved directly.
+* If you need to do that kind of thing to make your scripting make
+sense, do it in methods annotated with `@NonCPS`. These methods will
+be executed outside of the Workflow-specific interpreter, so can do
```

Comment:
I'd say `These methods will not be CPS transformed by the Workflow interpreter, so ...`, to be more precise.